### PR TITLE
Added support for nodes dialing back to cluster.

### LINKED ISSUE #803: Reverse tunnels for individual nodes?
Are there any plans for support of individual nodes creating reverse tunnels to a proxy server without creating a new cluster? We have a case where we would like to have a single node setup at multiple different sites, but currently it looks like we would need to configure a cluster for each site just for a single node to use a reverse tunnel.

### PR DESCRIPTION
**Pull request by russjones**
**Description**

Updated `services.ReverseTunnel` to support type (proxy or node). For proxy types, which represent trusted cluster connections, when a `services.ReverseTunnel` is created, it's created on the remote side with name `/reverseTunnels/example.com`. For node types, `services.ReverseTunnel` is created on the main side as `/reverseTunnels/{nodeUUID}.clusterName`.

Updated `services.TunnelConn` to support type (proxy or node). For proxy types, which represent trusted cluster connections, tunnel connections are created on the main side under `/tunnelConnections/remote.example.com/{proxyUUID}-remote.example.com`. For nodes, tunnel connections are created on the main side under `/tunnelConnections/example.com/{proxyUUID}-example.com`. This allows searching for tunnel connections by cluster then allows easily creating a set of proxies that are missing matching `services.TunnelConn`.

The reverse tunnel server has been updated to handle heartbeats from proxies as well as nodes. Proxy heartbeat behavior has not changed. Heartbeats from nodes now add remote connections to the matching local site. In addition, the reverse tunnel server now proxies connection to the Auth Server for requests that are already authenticated (a second authentication to the Auth Server is required).

For registration, nodes try and connect to the Auth Server to fetch host credentials. Upon failure, nodes now try and fallback to fetching host credentials from the web proxy.

To establish a connection to an Auth Server, nodes first try and connect directly, and if the connection fails, fallback to obtaining a connection to the Auth Server through the reverse tunnel. If a connection is established directly, node startup behavior has not changed. If a node establishes a connection through the reverse tunnel, it creates an AgentPool that attempts to dial back to the cluster and establish a reverse tunnel.

When nodes heartbeat, they also heartbeat if they are connected directly to the cluster or through a reverse tunnel. For nodes that are connected through a reverse tunnel, the proxy subsystem now directs the reverse tunnel server to establish a connection through the reverse tunnel instead of directly.

When sending discovery requests, the domain field has been replaced with tunnelID. The tunnelID field is either the cluster name (same as before) for proxies, or `{nodeUUID}.example.com` for nodes.

**Related Issues**

Fixes https://github.com/gravitational/teleport/issues/803

### CODE REVIEW

**Comment by klizhentas** [comment]
>  For node types, services.ReverseTunnel is created on the main side as /reverseTunnels/{nodeUUID}.clusterName

I'm pretty sure this not going to be scalable for 1K+ clusters

File: lib/service/connect.go [hunk 1]
```diff
@@ -742,7 +778,125 @@ func (process *TeleportProcess) rotate(conn *Connector, localState auth.StateV2,
...
+	}
+
+	tlsConfig, err := identity.TLSConfig(process.Config.CipherSuites)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	// Connect to the Auth Server through the reverse tunnel server.
+	authDialer := func(in context.Context, network, addr string) (net.Conn, error) {
+		channel, _, err := conn.OpenChannel(reversetunnel.ChanTransport, nil)
```
Comment (russjones) [inline lib/service/connect.go]: `reversetunnel.Dial`

File: lib/reversetunnel/srv.go [hunk 2]
```diff
@@ -490,26 +493,95 @@ func (s *server) HandleNewChan(conn net.Conn, sconn *ssh.ServerConn, nch ssh.New
...
+func (s *server) handleHeartbeat(conn net.Conn, sconn *ssh.ServerConn, nch ssh.NewChannel) {
 	s.Debugf("New tunnel from %v.", sconn.RemoteAddr())
 	if sconn.Permissions.Extensions[extCertType] != extCertTypeHost {
 		s.Error(trace.BadParameter("can't retrieve certificate type in certType"))
 		return
 	}
+
+	// Extract the role. For proxies, it's another cluster asking to join, for
+	// nodes it's a node dialing back.
+	val, ok := sconn.Permissions.Extensions["role"]
```
Comment (russjones) [inline lib/reversetunnel/srv.go]: constant

File: lib/reversetunnel/localsite.go [hunk 3]
```diff
@@ -216,18 +240,149 @@ func (s *localSite) dialWithAgent(params DialParams) (net.Conn, error) {
...
+		fmt.Sprintf("%v-%v", s.srv.ID, s.domainName),
+		services.TunnelConnectionSpecV2{
+			Type:          services.NodeTunnel,
+			ClusterName:   s.domainName,
+			ProxyName:     s.srv.ID,
+			LastHeartbeat: time.Now().UTC(),
+		},
+	)
+	tunnelConn.SetLastHeartbeat(t)
+	tunnelConn.SetExpiry(time.Now().Add(defaults.ReverseTunnelOfflineThreshold))
```
Comment (russjones) [inline lib/reversetunnel/localsite.go]: Use same delay as `services.ReverseTunnel`.

File: lib/reversetunnel/localsite.go [hunk 4]
```diff
@@ -216,18 +240,149 @@ func (s *localSite) dialWithAgent(params DialParams) (net.Conn, error) {
...
+	// Loop over all connections and remove and invalid connections from the
+	// connection map.
+	for key, _ := range s.remoteConns {
+		if s.remoteConns[key].isInvalid() {
+			delete(s.remoteConns, key)
 		}
 	}
-	return nil, trace.NotFound("server %v is unknown", addr)
+
+	rconn, ok := s.remoteConns[addr]
```
Comment (russjones) [inline lib/reversetunnel/localsite.go]: https://github.com/gravitational/teleport/blob/master/lib/reversetunnel/remotesite.go#L178

File: lib/reversetunnel/conn.go [hunk 5]
```diff
@@ -0,0 +1,500 @@
...
+		return
+	}
+
+	server := string(req.Payload)
+	var servers []string
+
+	// If the request is for the remote auth server or Kubernetes proxy, resolve
+	// and connect to them. Otherwise connect to the passed in server.
+	switch server {
+	case RemoteAuthServer:
```
Comment (russjones) [inline lib/reversetunnel/conn.go]: Comment.

File: lib/reversetunnel/conn.go [hunk 6]
```diff
@@ -0,0 +1,500 @@
...
+		fmt.Fprint(p.channel.Stderr(), err.Error())
+		req.Reply(false, []byte(err.Error()))
+		return
+	}
+
+	// Dail was successful.
+	req.Reply(true, []byte("Connected."))
+	p.log.Debugf("Successfully dialed to %v, start proxying.", server)
+
+	wg := sync.WaitGroup{}
```
Comment (russjones) [inline lib/reversetunnel/conn.go]: Context.

File: lib/reversetunnel/discovery.go [hunk 7]
```diff
@@ -11,10 +11,21 @@ import (
...
 
+// discoveryRequest is a request sent from a connected proxy with the missing proxies.
 type discoveryRequest struct {
-	ClusterName string            `json:"cluster_name"`
-	ClusterAddr utils.NetAddr     `json:"-"`
-	Proxies     []services.Server `json:"proxies"`
+	// TunnelID identifies who the tunnel is connected to. For trusted clusters,
+	// the TunnelID is the name of the remote cluster (like example.com). For
+	// nodes, it is the nodeID (like 4a050852-23b5-4d6d-a45f-bed02792d453.example.com).
+	TunnelID string `json:"cluster_name"`
```
Comment (russjones) [inline lib/reversetunnel/discovery.go]: Backward compat.

File: lib/srv/regular/proxy.go [hunk 8]
```diff
@@ -287,8 +287,15 @@ func (t *proxySubsys) proxyToHost(
...
 	// DNS resolvable.
 	var serverAddr string
+	var useTunnel bool
 	if server != nil {
 		serverAddr = server.GetAddr()
+
+		// Check if server requested to be connected to over a reverse tunnel.
+		if server.GetUseTunnel() {
+			useTunnel = true
+			serverAddr = fmt.Sprintf("%v.%v", server.GetName(), t.clusterName)
```
Comment (russjones) [inline lib/srv/regular/proxy.go]: See if this can be removed and figured out in the site.

File: lib/service/service.go [hunk 9]
```diff
@@ -1421,12 +1472,48 @@ func (process *TeleportProcess) initSSH() error {
...
 }
 
+func (process *TeleportProcess) upsertTunnelForever(conn *Connector) {
+	process.upsertTunnel(conn)
+
+	for {
+		select {
+		case <-process.ExitContext().Done():
+			return
+		case <-time.Tick(defaults.UpsertTunnelHeartbeatTick):
```
Comment (russjones) [inline lib/service/service.go]: Match rate.

**Comment by speeltronics** [comment]
@russjones  I compiled the branch rjones/iot5 on both my client and my server and I cannot get this to work. The way I was doing this by creating a cluster for each client is no longer working either. 
I can see the node on my server, but when I click to connect to it I just get:

disconnected
failed connecting to node MYNODE.
ERROR REPORT:
Original Error: *trace.ConnectionProblemError MYNODE is offline
Stack Trace:
        ~/go/src/github.com/gravitational/teleport/lib/reversetunnel/remotesite.go:556 github.com/gravitational/teleport/lib/reversetunnel.(*remoteSite).connThroughTunnel
        ~/go/src/github.com/gravitational/teleport/lib/reversetunnel/remotesite.go:482 github.com/gravitational/teleport/lib/reversetunnel.(*remoteSite).DialTCP
        ~/go/src/github.com/gravitational/teleport/lib/reversetunnel/remotesite.go:476 github.com/gravitational/teleport/lib/reversetunnel.(*remoteSite).Dial
        ~/go/src/github.com/gravitational/teleport/lib/srv/regular/proxy.go:320 github.com/gravitational/teleport/lib/srv/regular.(*proxySubsys).proxyToHost
        ~/go/src/github.com/gravitational/teleport/lib/srv/regular/proxy.go:190 github.com/gravitational/teleport/lib/srv/regular.(*proxySubsys).Start
        ~/go/src/github.com/gravitational/teleport/lib/srv/regular/sshserver.go:1145 github.com/gravitational/teleport/lib/srv/regular.(*Server).handleSubsystem
        ~/go/src/github.com/gravitational/teleport/lib/srv/regular/sshserver.go:1022 github.com/gravitational/teleport/lib/srv/regular.(*Server).dispatch
        ~/go/src/github.com/gravitational/teleport/lib/srv/regular/sshserver.go:987 github.com/gravitational/teleport/lib/srv/regular.(*Server).handleSessionRequests
        /usr/local/go/src/runtime/asm_amd64.s:2362 runtime.goexit
User Message: MYNODE is offline

Do you have example yaml files for the server and client or a cli I can use to test this? Thanks!

**Comment by speeltronics** [comment]
Could it be the roles to use the username on the client isn't working on this branch...

I'm getting this on the client.

Access to read role in namespace default denied to roles RemoteProxy,default-implicit-role: no allow rule matched. services/role.go:1720

**Comment by speeltronics** [comment]
I still cannot get this to work. If I connect a node to my my main node running the auth server it will show up on the web panel, but when I try to connect to it, I get connected to the main node running the auth server. The same thing happens whether my main auth server is runnning locally or remotely. I also tried specifying the actual node ip on the ssh instead of using 0.0.0.0, but then it timesout and never connects.

**Comment by russjones** [comment]
@gravitational-jenkins retest this please

**Pull request merged**